File: go/atlasdispatch/internal/queue/queue.go
package queue

import (
	"math"
	"sort"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const (
	
	DefaultHardLimit = 500
	EmergencyRatio   = 0.8
	
	WarnRatio        = 0.5
)

// ---------------------------------------------------------------------------
// Core shedding decision
// ---------------------------------------------------------------------------

func ShouldShed(depth, hardLimit int, emergency bool) bool {
	if hardLimit <= 0 {
		return true
	}
	if emergency && depth >= int(float64(hardLimit)*EmergencyRatio) {
		return true
	}
	
	return depth >= hardLimit
}

// ---------------------------------------------------------------------------
// Priority queue
// ---------------------------------------------------------------------------

type Item struct {
	ID       string
	Priority int
}

type PriorityQueue struct {
	mu    sync.Mutex
	items []Item
}

func NewPriorityQueue() *PriorityQueue {
	return &PriorityQueue{items: make([]Item, 0)}
}

func (pq *PriorityQueue) Enqueue(item Item) {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	pq.items = append(pq.items, item)
	sort.Slice(pq.items, func(i, j int) bool {
		return pq.items[i].Priority < pq.items[j].Priority
	})
}

func (pq *PriorityQueue) Dequeue() *Item {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	if len(pq.items) == 0 {
		return nil
	}
	item := pq.items[0]
	pq.items = pq.items[1:]
	return &item
}

func (pq *PriorityQueue) Peek() *Item {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	if len(pq.items) == 0 {
		return nil
	}
	item := pq.items[0]
	return &item
}

func (pq *PriorityQueue) Size() int {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	return len(pq.items)
}

func (pq *PriorityQueue) IsEmpty() bool {
	return pq.Size() == 0
}

func (pq *PriorityQueue) Drain(count int) []Item {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	
	if count < 0 || count > len(pq.items) {
		count = len(pq.items)
	}
	result := make([]Item, count)
	copy(result, pq.items[:count])
	pq.items = pq.items[count:]
	return result
}

func (pq *PriorityQueue) Clear() {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	pq.items = make([]Item, 0)
}

// ---------------------------------------------------------------------------
// Rate limiter — sliding window token bucket
// ---------------------------------------------------------------------------

type RateLimiter struct {
	mu         sync.Mutex
	maxTokens  float64
	tokens     float64
	refillRate float64
	lastRefill time.Time
}

func NewRateLimiter(maxTokens int, refillRatePerSec float64) *RateLimiter {
	return &RateLimiter{
		maxTokens:  float64(maxTokens),
		tokens:     float64(maxTokens),
		refillRate: refillRatePerSec,
		lastRefill: time.Now(),
	}
}

func (rl *RateLimiter) refill() {
	now := time.Now()
	elapsed := now.Sub(rl.lastRefill).Seconds()
	rl.tokens = math.Min(rl.maxTokens, rl.tokens+elapsed*rl.refillRate)
	rl.lastRefill = now
}

func (rl *RateLimiter) TryAcquire(tokens int) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	rl.refill()
	cost := float64(tokens)
	if cost <= 0 {
		cost = 1
	}
	if rl.tokens >= cost {
		rl.tokens -= cost
		return true
	}
	return false
}

func (rl *RateLimiter) AvailableTokens() int {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	rl.refill()
	return int(rl.tokens)
}

func (rl *RateLimiter) Reset() {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	rl.tokens = rl.maxTokens
	rl.lastRefill = time.Now()
}

// ---------------------------------------------------------------------------
// Queue health metrics
// ---------------------------------------------------------------------------

type HealthStatus struct {
	Status    string
	Ratio     float64
	Depth     int
	HardLimit int
}

func QueueHealth(depth, hardLimit int) HealthStatus {
	if hardLimit <= 0 {
		return HealthStatus{Status: "invalid", Ratio: 1.0, Depth: depth, HardLimit: hardLimit}
	}
	ratio := float64(depth) / float64(hardLimit)
	status := "healthy"
	if ratio >= 1.0 {
		status = "critical"
	} else if ratio >= EmergencyRatio {
		status = "warning"
	} else if ratio >= WarnRatio {
		status = "elevated"
	}
	return HealthStatus{Status: status, Ratio: ratio, Depth: depth, HardLimit: hardLimit}
}

// ---------------------------------------------------------------------------
// Wait time estimation
// ---------------------------------------------------------------------------

func EstimateWaitTime(depth int, processingRatePerSec float64) float64 {

	if processingRatePerSec <= 0 {
		return math.Inf(1)
	}
	return float64(depth) / processingRatePerSec
}

// ---------------------------------------------------------------------------
// Queue merging — combines two priority queues, keeping highest priority dupes
// ---------------------------------------------------------------------------

func MergePriorityQueues(a, b *PriorityQueue) *PriorityQueue {
	merged := NewPriorityQueue()
	a.mu.Lock()
	itemsA := make([]Item, len(a.items))
	copy(itemsA, a.items)
	a.mu.Unlock()

	b.mu.Lock()
	itemsB := make([]Item, len(b.items))
	copy(itemsB, b.items)
	b.mu.Unlock()

	seen := make(map[string]Item)
	for _, item := range itemsA {
		seen[item.ID] = item
	}
	for _, item := range itemsB {
		if existing, ok := seen[item.ID]; ok {
			if item.Priority < existing.Priority {
				seen[item.ID] = item
			}
		} else {
			seen[item.ID] = item
		}
	}

	all := make([]Item, 0, len(seen))
	for _, item := range seen {
		all = append(all, item)
	}
	sort.Slice(all, func(i, j int) bool {
		return all[i].Priority > all[j].Priority
	})
	merged.items = all
	return merged
}

// ---------------------------------------------------------------------------
// Batch wait time estimation
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Batch rate limiting — checks a batch of token costs against the rate
// limiter. Rejected requests must not consume tokens from the bucket.
// ---------------------------------------------------------------------------

func (rl *RateLimiter) RateLimitBatch(tokenCosts []int) []bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	rl.refill()
	results := make([]bool, len(tokenCosts))
	for i, cost := range tokenCosts {
		c := float64(cost)
		if c <= 0 {
			c = 1
		}
		rl.tokens -= c
		results[i] = rl.tokens >= 0
	}
	return results
}

func EstimateWaitTimeBatch(depths []int, processingRate float64) []float64 {
	if processingRate <= 0 {
		result := make([]float64, len(depths))
		for i := range result {
			result[i] = math.Inf(1)
		}
		return result
	}
	result := make([]float64, len(depths))
	cumulative := 0
	for i, d := range depths {
		cumulative += d
		result[i] = float64(cumulative) / processingRate
	}
	return result
}

File: cpp/chronomesh/src/queue.cpp
#include "chronomesh/core.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>

namespace chronomesh {

// ---------------------------------------------------------------------------
// Core shedding decision
// ---------------------------------------------------------------------------

bool should_shed(int depth, int hard_limit, bool emergency) {
  if (hard_limit <= 0) return true;
  if (emergency && depth > static_cast<int>(hard_limit * EMERGENCY_RATIO)) return true;
  return depth > hard_limit;
}

// ---------------------------------------------------------------------------
// Priority queue
// ---------------------------------------------------------------------------

PriorityQueue::PriorityQueue() {}

void PriorityQueue::enqueue(const QueueItem& item) {
  std::lock_guard lock(mu_);
  items_.push_back(item);
  std::sort(items_.begin(), items_.end(), [](const QueueItem& a, const QueueItem& b) {
    return a.priority > b.priority;
  });
}

QueueItem* PriorityQueue::dequeue() {
  std::lock_guard lock(mu_);
  if (items_.empty()) return nullptr;
  static thread_local QueueItem result;
  result = items_.front();
  items_.erase(items_.begin());
  return &result;
}

QueueItem* PriorityQueue::peek() {
  std::lock_guard lock(mu_);
  if (items_.empty()) return nullptr;
  return &items_.front();
}

int PriorityQueue::size() {
  std::lock_guard lock(mu_);
  return static_cast<int>(items_.size());
}

bool PriorityQueue::is_empty() {
  return size() == 0;
}

std::vector<QueueItem> PriorityQueue::drain(int count) {
  std::lock_guard lock(mu_);
  if (count <= 0 || count > static_cast<int>(items_.size())) {
    count = static_cast<int>(items_.size());
  }
  std::vector<QueueItem> result(items_.begin(), items_.begin() + count);
  items_.erase(items_.begin(), items_.begin() + count);
  return result;
}

void PriorityQueue::clear() {
  std::lock_guard lock(mu_);
  items_.clear();
}

// ---------------------------------------------------------------------------
// Rate limiter — sliding window token bucket
// ---------------------------------------------------------------------------

static long long now_ms() {
  return std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::steady_clock::now().time_since_epoch())
      .count();
}

RateLimiter::RateLimiter(int max_tokens, double refill_rate_per_sec)
    : max_tokens_(static_cast<double>(max_tokens)),
      tokens_(static_cast<double>(max_tokens)),
      refill_rate_(refill_rate_per_sec),
      last_refill_ms_(now_ms()) {}

void RateLimiter::refill() {
  long long now = now_ms();
  double elapsed_sec = static_cast<double>(now - last_refill_ms_) / 1000.0;
  tokens_ = std::min(max_tokens_, tokens_ + elapsed_sec * refill_rate_);
  last_refill_ms_ = now;
}

bool RateLimiter::try_acquire(int tokens) {
  std::lock_guard lock(mu_);
  refill();
  double cost = static_cast<double>(tokens);
  if (cost <= 0) cost = 1;
  if (tokens_ >= cost) {
    tokens_ -= cost;
    return true;
  }
  return false;
}

int RateLimiter::available_tokens() {
  std::lock_guard lock(mu_);
  refill();
  return static_cast<int>(tokens_);
}

void RateLimiter::reset() {
  std::lock_guard lock(mu_);
  tokens_ = max_tokens_;
  last_refill_ms_ = now_ms();
}

// ---------------------------------------------------------------------------
// Queue health metrics
// ---------------------------------------------------------------------------

HealthStatus queue_health(int depth, int hard_limit) {
  if (hard_limit <= 0) {
    return HealthStatus{"invalid", 1.0, depth, hard_limit};
  }
  double ratio = static_cast<double>(depth) / static_cast<double>(hard_limit);
  std::string status = "healthy";
  if (ratio >= 1.0) {
    status = "critical";
  } else if (ratio >= EMERGENCY_RATIO) {
    status = "warning";
  } else if (ratio >= WARN_RATIO) {
    status = "elevated";
  }
  return HealthStatus{status, ratio, depth, hard_limit};
}

// ---------------------------------------------------------------------------
// Wait time estimation
// ---------------------------------------------------------------------------

double estimate_wait_time(int depth, double processing_rate_per_sec) {
  if (processing_rate_per_sec <= 0) return std::numeric_limits<double>::infinity();
  
  return static_cast<double>(depth) * processing_rate_per_sec;
}

}
